api/lib/handler.js

```diff
@@ -0,0 +1,130 @@
+/** @import {Repository, Logger} from './typings.d.ts' */
+
+/**
+ * @param {object} args
+ * @param {Repository} args.repository
+ * @param {string} args.domain
+ * @param {Logger} args.logger
+ */
+export function createHandler ({ repository, domain, logger }) {
+  /**
+   * @param {import('node:http').IncomingMessage} req
+   * @param {import('node:http').ServerResponse} res
+   */
+  return (req, res) => {
+    const start = new Date()
+    logger.request(`${req.method} ${req.url} ...`)
+    handleRequest(req, res, { repository, domain, logger })
+      .catch(err => errorHandler(res, err, logger))
+      .then(() => {
+        logger.request(`${req.method} ${req.url} ${res.statusCode} (${new Date() - start}ms)`)
+      })
+  }
+}
+
+/**
+ * @param {import('node:http').IncomingMessage} req
+ * @param {import('node:http').ServerResponse} res
+ * @param {object} args
+ * @param {Repository} args.repository
+ * @param {string} args.domain
+ * @param {Logger} args.logger
+ */
+async function handleRequest (req, res, { repository, domain, logger }) {
+  if (req.headers.host.split(':')[0] !== domain) {
+    return redirect(req, res, `https://${domain}${req.url}`, 301)
+  }
+
+  // Caveat! `new URL('//foo', 'http://127.0.0.1')` would produce "http://foo/" - not what we want!
+  const { pathname, searchParams } = new URL(`http://127.0.0.1${req.url}`)
+  const segs = pathname.split('/').filter(Boolean)
+  // const url = `/${segs.join('/')}`
+
+  if (req.method !== 'GET') {
+    return notFound(res)
+  }
+
+  if (segs[0] === 'sample' && segs[1] && segs[2]) {
+    await samplePiecePayloadBlocks(req, res, repository, segs[1], segs[2], searchParams)
+  } else if (segs[0] === 'ingestion-status' && segs[1]) {
+    await getProviderIngestionStatus(req, res, repository, segs[1])
+  } else {
+    notFound(res)
+  }
+}
+
+/**
+ * @param {import('node:http').IncomingMessage} req
+ * @param {import('node:http').ServerResponse} res
+ * @param {Repository} repository
+ * @param {string} providerId
+ * @param {string} pieceCid
+ * @param {URLSearchParams} searchParams
+ */
+async function samplePiecePayloadBlocks (req, res, repository, providerId, pieceCid, searchParams) {
+  const payloadCids = await repository.getPiecePayloadBlocks(providerId, pieceCid)
+  const body = {}
+  if (payloadCids.length) {
+    body.samples = payloadCids.slice(0, 1)
+    res.setHeader('cache-control', `public, max-age=${24 * 3600 /* 24 hours */}, immutable`)
+  } else {
+    body.error = 'PROVIDER_OR_PIECE_NOT_FOUND'
+    res.setHeader('cache-control', `public, max-age=${60 /* 1min */}`)
+  }
+  return json(res, body)
+}
+
+/**
+ * @param {import('node:http').IncomingMessage} req
+ * @param {import('node:http').ServerResponse} res
+ * @param {Repository} repository
+ * @param {string} providerId
+ */
+async function getProviderIngestionStatus (req, res, repository, providerId) {
+  const walkerState = await repository.getWalkerState(providerId)
```

Comment:
btw, why did you call `repository` `repository`?

Reply:
Repository is a well known design pattern, see e.g. https://www.martinfowler.com/eaaCatalog/repository.html

I am happy to rename the workspace (component) to something else, see https://github.com/filecoin-station/piece-indexer/pull/18#discussion_r1771116511

Reply:
This is the first time I'm hearing about this. I'm fine with it if this is the correct term

Reply:
I guess I have never worked on enterprise applications (which this one also isn't). Let's see if others find this confusing too